### ISSUE TITLE
Integrate Differential Privacy

### ISSUE THREAD

**Issue by JulienVig** [epfml/disco#613]
Differential Privacy is implemented but never used. It seems that it was intentionally removed during a code refactoring (in [this PR](https://github.com/epfml/disco/pull/578/files#diff-bb5fd2b79ca040c5cdd1e0c4ec57de0ef4cfe3413a482d6c9b4b35a38b0e93e1), l.179 of decentralized/base.ts)

- [ ] Integrate DP again in the new architecture code
- [ ] Look into how DP is implemented. It seems it is epsilon-DP that is implemented, which has a very strict privacy bound at the cost of lower performance. Most implementations use (epsilon, delta)-DP for a better trade-off between privacy and utility. However, the latter needs to track the privacy budget delta for each user. This is what the MomentAccount tracks in [Deep learning with differential privacy](https://arxiv.org/abs/1607.00133).
- [ ] Create test for the differential privacy feature

**Comment by martinjaggi** [comment]
@ysfalh expressed some interest in this, as they have a [decentralized DP scheme](https://openreview.net/forum?id=5JrlywYHRi) which is easy to implement (only needs pairwise agreement on each edge, which we have anyway already from the secAgg implementation)

**Comment by ysfalh** [comment]
> @ysfalh expressed some interest in this, as they have a [decentralized DP scheme](https://openreview.net/forum?id=5JrlywYHRi) which is easy to implement (only needs pairwise agreement on each edge, which we have anyway already from the secAgg implementation)

Yes, I am happy to help with adapting our [code](https://github.com/elfirdoussilab1/DECOR).

**Comment by hleepfl** [comment]
Seems to have been implemented
Edit: my bad I didn't read well enough

**Comment by hleepfl** [comment]
Closing since it's *mostly* implemented. Will open a separate issue for keeping track of privacy budgeting.
